### ISSUE TITLE
Database setup error

### ISSUE THREAD

**Issue by mhyleung** [PedroMTQ/mantis#28]
Dear all

We have been trying to set up the Mantis database, but we have been encountering the following errors:

```bash
 python  mantis/  check_installation
Executing command:
 mantis/ check_installation
Checking installation
#  External data folders:
------------------------------------------
Default hmms folder:
/disk/rdisk08/zhiyshen/zhiyshen/tools/mantis/mantis/hmm/
Custom hmms folder:
/disk/rdisk08/zhiyshen/zhiyshen/tools/mantis/mantis/hmm/custom_hmms/
NCBI dump folder:
/disk/rdisk08/zhiyshen/zhiyshen/tools/mantis/mantis/Resources/NCBI/
Gene ontology folder:
/disk/rdisk08/zhiyshen/zhiyshen/tools/mantis/mantis/Resources/Gene_Ontology/
Uniprot dump folder:
/disk/rdisk08/zhiyshen/zhiyshen/tools/mantis/mantis/Resources/Uniprot/
TAX NOG hmms folder:
/disk/rdisk08/zhiyshen/zhiyshen/tools/mantis/mantis/hmm/NOGT/
Global NOG hmms folder:
/disk/rdisk08/zhiyshen/zhiyshen/tools/mantis/mantis/hmm/NOGG/
Pfam hmms folder:
/disk/rdisk08/zhiyshen/zhiyshen/tools/mantis/mantis/hmm/pfam/
KOfam hmms folder:
/disk/rdisk08/zhiyshen/zhiyshen/tools/mantis/mantis/hmm/kofam/
TIGRFAM hmms folder:
/disk/rdisk08/zhiyshen/zhiyshen/tools/mantis/mantis/hmm/tigrfam/
------------------------------------------
Cython correctly compiled!
Checking extra files
Checking extra files
Passed installation check on: /disk/rdisk08/zhiyshen/zhiyshen/tools/mantis/mantis/Resources/essential_genes
Passed installation check on: /disk/rdisk08/zhiyshen/zhiyshen/tools/mantis/mantis/Resources/NCBI/
Passed installation check on: /disk/rdisk08/zhiyshen/zhiyshen/tools/mantis/mantis/Resources/Uniprot/
Passed installation check on: /disk/rdisk08/zhiyshen/zhiyshen/tools/mantis/mantis/Resources/Gene_Ontology/
Checking HMM installation
Failed installation check on [files missing]: /disk/rdisk08/zhiyshen/zhiyshen/tools/mantis/mantis/hmm/NOGT/10/
10_sql_annotations.tsv
Failed installation check on [files missing]: /disk/rdisk08/zhiyshen/zhiyshen/tools/mantis/mantis/hmm/NOGT/104264/
104264_sql_annotations.tsv
Failed installation check on [files missing]: /disk/rdisk08/zhiyshen/zhiyshen/tools/mantis/mantis/hmm/NOGT/1028384/
.h3m; .h3p; .hmm; .h3i; .h3f; 1028384_sql_annotations.tsv
Failed installation check on [files missing]: /disk/rdisk08/zhiyshen/zhiyshen/tools/mantis/mantis/hmm/NOGT/10656/
10656_sql_annotations.tsv
Failed installation check on [files missing]: /disk/rdisk08/zhiyshen/zhiyshen/tools/mantis/mantis/hmm/NOGT/1161/
1161_sql_annotations.tsv
Failed installation check on [files missing]: /disk/rdisk08/zhiyshen/zhiyshen/tools/mantis/mantis/hmm/NOGT/119043/
119043_sql_annotations.tsv
................................

.h3m; .h3p; .hmm; 57723_sql_annotations.tsv; .h3i; .h3f
Failed installation check on [files missing]: /disk/rdisk08/zhiyshen/zhiyshen/tools/mantis/mantis/hmm/NOGT/5794/
.h3m; .h3p; 5794_sql_annotations.tsv; .hmm; .h3i; .h3f
Failed installation check on [files missing]: /disk/rdisk08/zhiyshen/zhiyshen/tools/mantis/mantis/hmm/NOGT/5796/
5796_sql_annotations.tsv; .h3m; .h3p; .hmm; .h3i; .h3f
Failed installation check on [files missing]: /disk/rdisk08/zhiyshen/zhiyshen/tools/mantis/mantis/hmm/NOGT/5809/
.h3m; .h3p; .hmm; .h3i; .h3f; 5809_sql_annotations.tsv
Failed installation check on [files missing]: /disk/rdisk08/zhiyshen/zhiyshen/tools/mantis/mantis/hmm/NOGT/5819/
.h3m; .h3p; .hmm; .h3i; .h3f; 5819_sql_annotations.tsv
Failed installation check on [files missing]: /disk/rdisk08/zhiyshen/zhiyshen/tools/mantis/mantis/hmm/NOGT/583/
583_sql_annotations.tsv
Failed installation check on [files missing]: /disk/rdisk08/zhiyshen/zhiyshen/tools/mantis/mantis/hmm/NOGT/586/
586_sql_annotations.tsv
Failed installation check on [files missing]: /disk/rdisk08/zhiyshen/zhiyshen/tools/mantis/mantis/hmm/NOGT/5863/
5863_sql_annotations.tsv
Failed installation check on [files missing]: /disk/rdisk08/zhiyshen/zhiyshen/tools/mantis/mantis/hmm/NOGT/5878/
5878_sql_annotations.tsv; .h3m; .h3p; .hmm; .h3i; .h3f
Failed installation check on [files missing]: /disk/rdisk08/zhiyshen/zhiyshen/tools/mantis/mantis/hmm/NOGT/58840/
58840_sql_annotations.tsv
Failed installation check on [files missing]: /disk/rdisk08/zhiyshen/zhiyshen/tools/mantis/mantis/hmm/NOGT/590/
590_sql_annotations.tsv
Failed installation check on [files missing]: /disk/rdisk08/zhiyshen/zhiyshen/tools/mantis/mantis/hmm/NOGT/59732/
.h3m; .h3p; .hmm; .h3i; .h3f; 59732_sql_annotations.tsv
Failed installation check on [files missing]: /disk/rdisk08/zhiyshen/zhiyshen/tools/mantis/mantis/hmm/NOGT/60136/
.h3m; .h3p; .hmm; .h3i; .h3f; 60136_sql_annotations.tsv
Failed installation check on [files missing]: /disk/rdisk08/zhiyshen/zhiyshen/tools/mantis/mantis/hmm/NOGT/613/
.h3m; .h3p; .hmm; .h3i; .h3f; 613_sql_annotations.tsv
Failed installation check on [files missing]: /disk/rdisk08/zhiyshen/zhiyshen/tools/mantis/mantis/hmm/NOGT/61432/
61432_sql_annotations.tsv
Failed installation check on [files missing]: /disk/rdisk08/zhiyshen/zhiyshen/tools/mantis/mantis/hmm/NOGT/622450/
622450_sql_annotations.tsv; .h3m; .h3p; .hmm; .h3i; .h3f
Failed installation check on [files missing]: /disk/rdisk08/zhiyshen/zhiyshen/tools/mantis/mantis/hmm/NOGT/6231/
6231_sql_annotations.tsv; .h3m; .h3p; .hmm; .h3i; .h3f
Failed installation check on [files missing]: /disk/rdisk08/zhiyshen/zhiyshen/tools/mantis/mantis/hmm/NOGT/6236/
.h3m; .h3p; .hmm; 6236_sql_annotations.tsv; .h3i; .h3f
Failed installation check on [files missing]: /disk/rdisk08/zhiyshen/zhiyshen/tools/mantis/mantis/hmm/NOGT/629/
.h3m; .h3p; .hmm; .h3i; .h3f; 629_sql_annotations.tsv
Failed installation check on [files missing]: /disk/rdisk08/zhiyshen/zhiyshen/tools/mantis/mantis/hmm/NOGT/629295/
629295_sql_annotations.tsv; .h3m; .h3p; .hmm; .h3i; .h3f
Failed installation check on [files missing]: /disk/rdisk08/zhiyshen/zhiyshen/tools/mantis/mantis/hmm/NOGT/639021/
639021_sql_annotations.tsv; .h3m; .h3p; .hmm; .h3i; .h3f
Failed installation check on [files missing]: /disk/rdisk08/zhiyshen/zhiyshen/tools/mantis/mantis/hmm/NOGT/651137/
651137_sql_annotations.tsv; .h3m; .h3p; .hmm; .h3i; .h3f
Failed installation check on [files missing]: /disk/rdisk08/zhiyshen/zhiyshen/tools/mantis/mantis/hmm/NOGT/6656/
.h3m; .h3p; .hmm; .h3i; .h3f; 6656_sql_annotations.tsv
Failed installation check on [files missing]: /disk/rdisk08/zhiyshen/zhiyshen/tools/mantis/mantis/hmm/NOGT/671232/
.h3m; .h3p; .hmm; .h3i; 671232_sql_annotations.tsv; .h3f
Failed installation check on [files missing]: /disk/rdisk08/zhiyshen/zhiyshen/tools/mantis/mantis/hmm/NOGT/675063/
675063_sql_annotations.tsv; .h3m; .h3p; .hmm; .h3i; .h3f
Failed installation check on [files missing]: /disk/rdisk08/zhiyshen/zhiyshen/tools/mantis/mantis/hmm/NOGG/
.h3m; .h3p; .hmm; .h3i; NOGG_sql_annotations.tsv; .h3f
Passed installation check on: /disk/rdisk08/zhiyshen/zhiyshen/tools/mantis/mantis/hmm/pfam/
Passed installation check on: /disk/rdisk08/zhiyshen/zhiyshen/tools/mantis/mantis/hmm/kofam/
Passed installation check on: /disk/rdisk08/zhiyshen/zhiyshen/tools/mantis/mantis/hmm/tigrfam/
Installation check failed on: 303
------------------------------------------
--------INSTALLATION CHECK FAILED!--------
------------------------------------------
```

I am not entire sure why some of the files are missing in these NOGG/NOGT subdirectories. If anyone can point me in the right direction that would be great! Thanks

Regards

Marc

**Comment by PedroMTQ** [comment]
Hello Marc,

Could you please let me know which version of Mantis you are currently using? The latest release no longer uses the NOGG or NOGT folder structure.
In general, the NOG folders (e.g.,`/disk/rdisk08/zhiyshen/zhiyshen/tools/mantis/mantis/hmm/NOGT/2/`) should contain:
![image](https://user-images.githubusercontent.com/22548082/129901642-577b6e0c-8f2e-425d-ba60-d82efeacabf7.png)

Could you let me know what files your folder(s) contains?
(The chunks folder is only present for larger HMMs)

Regards,
Pedro

**Comment by mhyleung** [comment]
Hi Pedro thanks for the quick response.

We are using version 1.1, and we have about 700GB of files on the database, but I have a feeling that not all of the files have been downloaded successfully, and we do not know how much we are missing.

Would you recommend us starting the download again, and use the --chunk-size option to break the downloads into smaller bits?

Thanks

Marc

**Comment by PedroMTQ** [comment]
Hi Marc,
Please use v1.1.1 as it includes some structure changes.
Yes, I believe part of the database is missing, my local copy for the NOG database is around 3TB.
As you said, you should restart the download of the NOG database, but all others seem fine. To do that you can run `setup_databases` again; I would probably recommend you remove your old NOG folder though, just to make sure everything is correctly structured.

Regarding the chunk_size, that is only used for splitting the HMMs and samples. At the moment, we only download via streamed chunks if the normal downloading fails (you should get a warning if the download fails).

Keep me updated.

Regards,
Pedro

**Comment by mhyleung** [comment]
Hi Pedro

My colleague has tried to set up v1.1.1, and tried downloading the databases again. He encountered the following errors:

```bash
python mantis -cs 5000 setup_databases

zhiyshen 408025  0.0  0.0 2743476  732 ?        Sl   Aug27   1:19 python mantis -cs 5000 setup_databases
zhiyshen 409630  0.0  0.0 407760    56 ?        S    Aug27   0:03 python mantis -cs 5000 setup_databases

shen/tools/mantis/mantis/References/NOG/1305/profiles
Process Process-4:
Traceback (most recent call last):
  File "/home/zhiyshen/anaconda2/envs/mantis_env/lib/python3.8/multiprocessing/process.py", line 315, in _bootstrap
    self.run()
  File "/home/zhiyshen/anaconda2/envs/mantis_env/lib/python3.8/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "mantis/source/MANTIS_DB.py", line 183, in worker_setup_databases
    self.download_database(database, force_download, taxon_id, stdout_path)
  File "mantis/source/MANTIS_DB.py", line 218, in download_database
    elif database == 'kofam':       self.download_kofam(force_download=force_download, stdout_file=stdout_file)
  File "mantis/source/MANTIS_DB.py", line 546, in download_kofam
    self.compile_kofam_metadata()
  File "mantis/source/MANTIS_DB.py", line 412, in compile_kofam_metadata
    self.get_link_kofam_ko_to_pathway(metadata_to_write)
  File "mantis/source/MANTIS_DB.py", line 462, in get_link_kofam_ko_to_pathway
    map_description = self.get_kofam_pathway_description()
  File "mantis/source/MANTIS_DB.py", line 513, in get_kofam_pathway_description
    res[kegg_map] = {'description': description, 'parent': sub_title, 'grand_parent': main_tile}
UnboundLocalError: local variable 'main_tile' referenced before assignment

.....
Exception
Did not manage to download the following url correctly:
http://eggnog5.embl.de/download/latest/per_tax_level/186801/186801_hmms.tar.gz
Process Process-20:
Traceback (most recent call last):
  File "/home/zhiyshen/anaconda2/envs/mantis_env/lib/python3.8/multiprocessing/process.py", line 315, in _bootstrap
    self.run()
  File "/home/zhiyshen/anaconda2/envs/mantis_env/lib/python3.8/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "mantis/source/MANTIS_DB.py", line 183, in worker_setup_databases
    self.download_database(database, force_download, taxon_id, stdout_path)
  File "mantis/source/MANTIS_DB.py", line 222, in download_database
    elif database == 'NOG':         self.download_NOGT(taxon_id=taxon_id, stdout_file=stdout_file)
  File "mantis/source/MANTIS_DB.py", line 936, in download_NOGT
    download_file(url, output_folder=folder_path, stdout_file=stdout_file)
  File "mantis/source/utils.py", line 651, in download_file
    raise Exception
Exception
```

Thank you

Regards

Marcus

**Comment by PedroMTQ** [comment]
Dear Marcus,

It seems you have two unrelated errors, from what I can see the first one is regarding kofam metadata. I've now included a fix for this bug.
The second error seems to be an issue downloading the eggNOG database, another user had a similar [issue](https://github.com/PedroMTQ/mantis/issues/15).

Please try the new release 1.2, it should address the first error. Regarding the second error, could you please include a full traceback?

Regards,
Pedro

**Comment by mhyleung** [comment]
Hi Pedro

Thank you for your email. my colleague has updated to v1.2 and attached the log file for the database download step. Thank you for your help.

Marcus
[nohup.txt](https://github.com/PedroMTQ/mantis/files/7128128/nohup.txt)

**Comment by PedroMTQ** [comment]
Hello Marcus,

Thank you for the log file. It seems the link I previously used to download the TIGRfams data is no longer available. However this is not an issue as these HMMs are already contained within the NCBI HMMs. I've published release 1.2.1 with this bug fix so please download it.

There is also an issue with the eggNOG HMMs download, unfortunately this seems to be beyond my control. It seems some HMMs are not being download even though they are available. I've tested downloading (with Mantis) some of the HMMs that appeared to crash in your log file and it worked.
The eggNOG HMMs tend to be quite large, so it might be space restrictions? Or maybe there are some other restrictions in place regarding consecutive or streamed downloads? I'm not entirely sure. One thing you could do is manually download the HMMs yourself with the same structure required by Mantis, i.e.:
NOG/**taxa_id**/**taxa_id**_annotations.tsv.gz
NOG/**taxa_id**/**taxa_id**_.hmms.tar.gz

Then just comment (add `#` before the line) the line: 
`download_file(url, output_folder=folder_path, stdout_file=stdout_file)` in the function `download_NOGT`

You can then rerun `setup_databases` and, provided you downloaded the files and formatted the folders correctly, and everything should work.

**Comment by PedroMTQ** [comment]
@mhyleung  I'm closing this issue. Please re-open if it has not been solved.

Regards,
Pedro
